src/mlpack/core/optimizers/spsa/spsa.hpp

```diff
@@ -0,0 +1,117 @@
+/**
+ * @file spsa_update.hpp
+ * @author N Rajiv Vaidyanathan
+ *
+ * SPSA update for faster convergence.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_SPSA_SPSA_HPP
+#define MLPACK_CORE_OPTIMIZERS_SPSA_SPSA_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/optimizers/sgd/sgd.hpp>
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Implementation of the SPSA update policy. SPSA update policy improves
+ * the rate of optimization by performing simultaneously on all parameters.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{Spall1998,
+ *   author  = {Spall, J. C.},
+ *   title   = {An Overview of the Simultaneous Perturbation Method for
+ *              Efficient Optimization},
+ *   journal = {Johns Hopkins APL Technical Digest},
+ *   volume  = {19},
+ *   number  = {4},
+ *   pages   = {482--492},
+ *   year    = {1998}
+ * }
+ * @endcode
+ *
+ */
+class SPSA
+{
+ public:
+  // Specifies the maximum number of iterations.
+  long long int max_iter;
+  arma::vec sp_vector;
+
+  // Control the amount of gradient update.
+  float a;
+  float A;
+
+  // Non-negative co-efficients controlling the optimizer.
+  float alpha;
```

Comment:
Usually we provide separated functions to modify the method parameter, see https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/optimizers/sgd/sgd.hpp for an example.